test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,394 @@
+
+    cset_volume_diff1 "$CAP_CHANNELS"
+    wait
+
+    cset_nom
+    wait
+
+    cset_mute
+    wait
+
+    cset_unmute
+    wait
+
+    cset_mute
+    wait
+
+    cset_max
+    wait
+
+    cset_unmute
+    wait
+
+    cset_mutevol
+    wait
+
+    cset_min
+    wait
+
+    cset_volume_diff2 "$CAP_CHANNELS"
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 previous gains
+# 1..2s channels 1-4 different mute switches
+# 2..3s channels 1-4 all muted
+run_test_2 () {
+    arecord_in_background "$ARECORD_WAV2" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
+
+    cset_nom
+    cset_mute_diff1 "$CAP_CHANNELS"
+    wait
+
+    cset_mute
+    wait
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 muted
+# 1..2s channels 1-4  nominal gain
+# 2..3s different mute switches
+# 3..4s channels 1-4 nominal gain
+run_test_3 () {
+    arecord_in_background "$ARECORD_WAV3" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
+
+    cset_nom
+    wait
+
+    cset_mute_diff2 "$CAP_CHANNELS"
+    wait
+
+    cset_unmute
+
+    wait_for_pid "$PID"
+    dlogi Ready.
+}
+
+#
+# Helper functions
+#
+
+cset_nom () {
+    dlogi Set nominal volume
+    amixer cset name="$CAP_VOLUME" $VOL_NOM
+}
+
+cset_max () {
+    dlogi Set maximum volume
+    amixer cset name="$CAP_VOLUME" $VOL_MAX
+}
+
+cset_min () {
+    dlogi Set minimum volume
+    amixer cset name="$CAP_VOLUME" $VOL_MIN
+}
+
+cset_mutevol () {
+    dlogi Set mute via volume
+    amixer cset name="$CAP_VOLUME" $VOL_MUTE
+}
+cset_mute () {
+    dlogi Set mute via switch
+    amixer cset name="$CAP_SWITCH" off
+}
+
+cset_unmute () {
+    dlogi Set unmute
+    amixer cset name="$CAP_SWITCH" on
+}
+
+wait () {
+    dlogi Waiting
+    sleep 1
+}
+
+cset_mute_diff1 () {
+    dlogi Set switch pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" off,on,on,off
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" off,on
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" off
+	    ;;
+    esac
+}
+
+cset_mute_diff2 () {
+    dlogi Set switch pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" on,off,off,on
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" on,off
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" on
+	    ;;
+    esac
+}
+
+cset_volume_diff1 () {
+    dlogi Set volume pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB,$VOL_M10DB,$VOL_M30DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB
+	    ;;
+    esac
+}
+
+cset_volume_diff2 () {
+    dlogi Set volume pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB,$VOL_0DB,$VOL_M20DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB
+	    ;;
+    esac
+}
+
+wait_for_pid () {
+    dlogi Waiting for process "$1" to complete
+    tail --pid="$pid" -f /dev/null
+}
```

Comment:
Using `tail` like this is really a hack. It also fails to catch failures which is really bad for test code.

There's a clean way to do this:
```shell
timeout -k 5 10 some_background_ process & somePID=$!
...

wait_for_pid()
{
ret=0
wait $somePID || ret=$?
if [ $ret -eq 124 ] || [ $ret -eq 137 ]; then
 printf 'process %d didnt stop, we timed it out\n' "$somePID"
elif [ $ret -ne 0 ]
 printf 'process %d failed with status %d\n' "$somePID" "$ret"
fi
return $ret
}
```
See `man timeout`. Feel free to add a function like this to `case-lib/lib.sh`, that logic should be in use all across the tests and I don't think it is right now.

Reply:
Thanks, I'll try. Hope it solves the "aplay &" vs. "set -e" conflict.

Reply:
No success. Nohup might be an alternative too. I'm now upgrading Ubuntu in the test device in hope that my issue would be a bug in bash.

Reply:
Or put bash function background instead of aplay?

Reply:
I removed this function totally with use of wait %N. If %N is not reliable I need to return this this kind of function for PID. I'll test more the robustness with killed processes to see what happens.